README.md
SwissPaymentSlipFpdf - Swiss Payment Slips with FPDF
====================================================

[![Travis Build Status](https://travis-ci.org/ravage84/SwissPaymentSlipFpdf.png?branch=master)](https://travis-ci.org/ravage84/SwissPaymentSlipFpdf)
[![Scrutinizer CI Build Status](https://scrutinizer-ci.com/g/ravage84/SwissPaymentSlipFpdf/badges/build.png?b=master)](https://scrutinizer-ci.com/g/ravage84/SwissPaymentSlipFpdf/build-status/master)
[![Scrutinizer CI Code Coverage](https://scrutinizer-ci.com/g/ravage84/SwissPaymentSlipFpdf/badges/coverage.png?b=master)](https://scrutinizer-ci.com/g/ravage84/SwissPaymentSlipFpdf/?branch=master)
[![Scrutinizer CI Code Quality](https://scrutinizer-ci.com/g/ravage84/SwissPaymentSlipFpdf/badges/quality-score.png?b=master)](https://scrutinizer-ci.com/g/ravage84/SwissPaymentSlipFpdf/?branch=master)
[![Total Downloads](https://poser.pugx.org/swiss-payment-slip/swiss-payment-slip-fpdf/d/total.png)](https://packagist.org/packages/swiss-payment-slip/swiss-payment-slip-fpdf)
[![Latest Stable Version](https://poser.pugx.org/swiss-payment-slip/swiss-payment-slip-fpdf/v/stable.png)](https://packagist.org/packages/swiss-payment-slip/swiss-payment-slip-fpdf)

Do you need to create Swiss payment slips (called ESR) as PDF files in a project of yours?
Then you found the right place, read on...

By the way if you don't like [FPDF](http://www.fpdf.org) or don't want to use it for whatever reason,
checkout [SwissPaymentSlipTcpdf](https://github.com/ravage84/SwissPaymentSlipTcpdf/).

How to use
----------

Just install the package (see [Installation](https://github.com/ravage84/SwissPaymentSlipFcpdf#installation))
and check out the contained examples in the
[examples folder](https://github.com/ravage84/SwissPaymentSlipFcpdf/tree/master/examples).

How to extend for custom needs
------------------------------

[TODO]
If you need help, ask for help.

Installation
------------

### Requirements

- PHP 5.4.x+
- [SwissPaymentSlip](https://github.com/ravage84/SwissPaymentSlip/) (automatically installed by Composer)
- [SwissPaymentSlipPdf](https://github.com/ravage84/SwissPaymentSlipPdf/) (automatically installed by Composer)

### Composer

Just [install composer](http://getcomposer.org/doc/00-intro.md#system-requirements) on your system, if not already there.
Then create a [composer.json](http://getcomposer.org/doc/04-schema.md) file in your project's root folder
and copy the following into it:

```JSON
{
    "require": {
        "swiss-payment-slip/swiss-payment-slip-fpdf": "*"
    }
}
```

After that you can install the package using

    $ php composer.phar install

in your project's root folder.

Background Story
----------------

In february 2013 I was looking for a solution to create swiss payment slips for a project I had to do at my work place.
After a short Google search I came across Manuel Reinhard's
[blog post](http://sprain.ch/blog/downloads/class-esr-besr-einzahlungsschein-php/) about the class he made for that.
On his [Github project's page](https://github.com/sprain/class.Einzahlungsschein.php) I found
[Peter Siska's](https://github.com/peschee) [pull request](https://github.com/sprain/class.Einzahlungsschein.php/pull/5).
His pull request introduced PSR-0 compatibility and he created a composer package on [Packagist](http://packagist.org/).
So I tried Peter's version and it suited my basic needs.

BUT since the customer I was working for used custom designed payment slips I couldn't use Manuel's/Peter's script
since it wasn't flexible enough.
Now I had to decide whether I want to "just" change the script to fit my needs or to rewrite it and
make it as flexible as possible.
I decided myself for the latter.

TODOs
-----

- Improve code documentation
- Add more examples/improve existing ones
- Write tests
- Release the stable release of the API

Submitting bugs and feature requests
------------------------------------

Bugs and feature request are tracked on [GitHub](https://github.com/ravage84/SwissPaymentSlipFpdf/issues).

Author
------

This project was created by [Marc Würth](https://github.com/ravage84).
See Background Story for more details.

License
-------

SwissPaymentSlipFpdf is licensed under the MIT License.
See the [LICENSE](https://github.com/ravage84/SwissPaymentSlipFpdf/blob/master/LICENSE) file for details.

Thanks to
---------

- <http://ansuz.sooke.bc.ca/software/ocrb.php> for OCRB font
- <http://blog.fruit-lab.de/fpdf-font-converter/> for FPDF font converter


CHANGELOG.md
# Change Log
All notable changes to this project are documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased](https://github.com/ravage84/SwissPaymentSlipFpdf/compare/0.6.0...master)
### Added

### Changed
- removed support information for red payment slip (ES) #12

### Fixed

## [0.6.0](https://github.com/ravage84/SwissPaymentSlipFpdf/releases/tag/0.6.0) - 2015-04-01
### Added
- Scrutinizer CI integration & badges
- .editorconfig file
- PHPUnit 3.7.38 as development dependency
- Testing with newer PHP versions and HHVM through Travis CI
- A .gitattributes
- PHPCS 2.1.* as development dependency
- composer.lock (not ignored anymore)
- This change log
- Packagist Download & Latest badges to the README
- A bunch of incomplete tests

### Changed
- Updated the swiss-payment-slip/swiss-payment-slip-pdf dependency to version 0.13.* (API breaking)
- Adjusted the calls to the updated SwissPaymentSlipPdf class
- Added the PDFs generated through the examples to the ignore list
- Changed convertColor2Rgb method to protected instead of private
- Remove set minimum-stability ("dev"), get stable version of dependencies
- Set itbz/fpdf dependency to 1.7.*
- Fully adopted the PSR2 Code Style
- Renamed SwissPaymentSlipFpdf to PaymentSlipFpdf
- Adopted the PSR-4 autoloader standard
- Improved CS, DobBlocks and documentation
- Exclude development/testing only related stuff from the Composer package
- Implement a fluent interface

### Fixed
- Removed misleading time key, which fooled Packagist

## [0.5.0](https://github.com/ravage84/SwissPaymentSlipFpdf/releases/tag/0.5.0) - 2013-03-07
# Added
- Initial release including code, FPDF fonts, examples, a readme and a Travis configuration


tests/bootstrap.php
<?php
/**
 * Swiss Payment Slip FPDF
 *
 * @license http://www.opensource.org/licenses/mit-license.php MIT License
 * @copyright 2012-2015 Some nice Swiss guys
 * @author Marc Würth <ravage@bluewin.ch>
 * @author Manuel Reinhard <manu@sprain.ch>
 * @author Peter Siska <pesche@gridonic.ch>
 * @link https://github.com/ravage84/SwissPaymentSlipFpdf
 */

namespace SwissPaymentSlip\SwissPaymentSlipFpdf\Tests;

use SwissPaymentSlip\SwissPaymentSlip\PaymentSlip;
use SwissPaymentSlip\SwissPaymentSlip\PaymentSlipData;

// Include Composer's autoloader
require __DIR__.'/../vendor/autoload.php';

/**
 * A wrapping class to allow testing the abstract class PaymentSlipData
 */
class TestablePaymentSlipData extends PaymentSlipData
{
}

/**
 * A wrapping class to allow testing the abstract class PaymentSlip
 */
class TestablePaymentSlip extends PaymentSlip
{
}


tests/PaymentSlipFpdfTest.php
<?php
/**
 * Swiss Payment Slip FPDF
 *
 * @license http://www.opensource.org/licenses/mit-license.php MIT License
 * @copyright 2012-2015 Some nice Swiss guys
 * @author Marc Würth <ravage@bluewin.ch>
 * @author Manuel Reinhard <manu@sprain.ch>
 * @author Peter Siska <pesche@gridonic.ch>
 * @link https://github.com/ravage84/SwissPaymentSlipFpdf
 */

namespace SwissPaymentSlip\SwissPaymentSlipFpdf\Tests;

/**
 * Tests for the PaymentSlipFpdf class
 *
 * @coversDefaultClass SwissPaymentSlip\SwissPaymentSlipPdf\PaymentSlipFpdf
 */
class PaymentSlipFpdfTest extends \PHPUnit_Framework_TestCase
{
    /**
     * Tests the displayImage method
     *
     * @return void
     * @covers ::displayImage
     */
    public function testDisplayImage()
    {
        $this->markTestIncomplete(
            'This test has not been implemented yet.'
        );
    }

    /**
     * Tests the setFont method
     *
     * @return void
     * @covers ::setFont
     */
    public function testSetFont()
    {
        $this->markTestIncomplete(
            'This test has not been implemented yet.'
        );
    }

    /**
     * Tests the setBackground method
     *
     * @return void
     * @covers ::setBackground
     */
    public function testSetBackground()
    {
        $this->markTestIncomplete(
            'This test has not been implemented yet.'
        );
    }

    /**
     * Tests the setPosition method
     *
     * @return void
     * @covers ::setPosition
     */
    public function testSetPosition()
    {
        $this->markTestIncomplete(
            'This test has not been implemented yet.'
        );
    }

    /**
     * Tests the createCell method
     *
     * @return void
     * @covers ::createCell
     */
    public function testCreateCell()
    {
        $this->markTestIncomplete(
            'This test has not been implemented yet.'
        );
    }

    /**
     * Tests the convertColor2Rgb method
     *
     * @return void
     * @covers ::convertColor2Rgb
     * @covers ::hex2RGB
     */
    public function testConvertColor2Rgb()
    {
        $this->markTestIncomplete(
            'This test has not been implemented yet.'
        );
    }
}


examples/index.php
<?php
/**
 * Swiss Payment Slip FPDF
 *
 * @license http://www.opensource.org/licenses/mit-license.php MIT License
 * @copyright 2012-2015 Some nice Swiss guys
 * @author Marc Würth <ravage@bluewin.ch>
 * @author Manuel Reinhard <manu@sprain.ch>
 * @author Peter Siska <pesche@gridonic.ch>
 * @link https://github.com/ravage84/SwissPaymentSlipFpdf
 */
?>
<!DOCTYPE html>
<html>
<head>
    <meta charset="utf-8">
    <title>Examples for SwissPaymentSlipFpdf</title>
</head>
<body>
<h1>Examples for SwissPaymentSlipFpdf</h1>

<p>Learn how to use the SwissPaymentSlipFpdf class with the following examples:</p>
<ul>
    <li>
        <a href="SwissPaymentSlipFpdf_Orange_Slip.php">
            01 - Create an orange payment slip with SwissPaymentSlipFpdf
        </a>
    </li>
    <li>
        <a href="SwissPaymentSlipFpdf_Thousand_Slips.php">
            03 - Create one thousand orange payment slips with SwissPaymentSlipFpdf
        </a>
    </li>
</ul>
</body>
</html>


examples/SwissPaymentSlipFpdf_Red_Slip.php
<?php
/**
 * Swiss Payment Slip FPDF
 *
 * @license http://www.opensource.org/licenses/mit-license.php MIT License
 * @copyright 2012-2015 Some nice Swiss guys
 * @author Marc Würth <ravage@bluewin.ch>
 * @author Manuel Reinhard <manu@sprain.ch>
 * @author Peter Siska <pesche@gridonic.ch>
 * @link https://github.com/ravage84/SwissPaymentSlipFpdf
 */
?>
<!DOCTYPE html>
<html>
<head>
    <meta charset="utf-8">
    <title>SwissPaymentSlipFpdf Example 02: Create a red payment slip</title>
</head>
<body>
<h1>SwissPaymentSlipFpdf Example 02: Create a red payment slip</h1>
<?php
// Measure script execution/generating time
$time_start = microtime(true);

// Make sure the classes get auto-loaded
require __DIR__ . '/../vendor/autoload.php';

// Import necessary classes
use SwissPaymentSlip\SwissPaymentSlip\RedPaymentSlipData;
use SwissPaymentSlip\SwissPaymentSlip\RedPaymentSlip;
use SwissPaymentSlip\SwissPaymentSlipFpdf\PaymentSlipFpdf;
use fpdf\FPDF;

// Make sure FPDF has access to the additional fonts
define('FPDF_FONTPATH', __DIR__ . '/../src/Resources/font');

// Create an instance of FPDF, setup default settings
$fPdf = new FPDF('P', 'mm', 'A4');

// Add OCRB font to FPDF
$fPdf->AddFont('OCRB10');

// Add page, don't break page automatically
$fPdf->AddPage();
$fPdf->SetAutoPageBreak(false);

// Insert a dummy invoice text, not part of the payment slip itself
$fPdf->SetFont('Helvetica', '', 9);
$fPdf->Cell(50, 4, "Just some dummy text.");

// Create a payment slip data container (value object)
$paymentSlipData = new RedPaymentSlipData();

// Fill the data container with your data
$paymentSlipData->setBankData('Seldwyla Bank', '8021 Zürich');
$paymentSlipData->setAccountNumber('80-939-3');
$paymentSlipData->setRecipientData('Muster AG', 'Bahnhofstrasse 5', '8001 Zürich');
$paymentSlipData->setIban('CH3808888123456789012');
$paymentSlipData->setPayerData('M. Beispieler', 'Bahnhofstrasse 356', '', '7000 Chur');
$paymentSlipData->setAmount(8479.25);
$paymentSlipData->setPaymentReasonData('Rechnung', 'Nr.7496');

// Create a payment slip object, pass in the prepared data container
$paymentSlip = new RedPaymentSlip($paymentSlipData, 0, 191);

// Create an instance of the FPDF implementation
$paymentSlipFpdf = new PaymentSlipFpdf($fPdf, $paymentSlip);

// "Print" the slip with its elements according to their attributes
$paymentSlipFpdf->createPaymentSlip($paymentSlip);

// Output PDF named example_fpdf_red_slip.pdf to examples folder
$pdfName = 'example_fpdf_red_slip.pdf';
$pdfPath = __DIR__ . DIRECTORY_SEPARATOR . $pdfName;
$fPdf->Output($pdfPath, 'F');

echo sprintf('Payment slip created in <a href="%s">%s</a><br>', $pdfName, $pdfPath);

echo "<br>";

$time_end = microtime(true);
$time = $time_end - $time_start;
echo "Generation took $time seconds <br>";
echo 'Peak memory usage: ' . memory_get_peak_usage() / 1024 / 1024;
?>
</body>
</html>


examples/SwissPaymentSlipFpdf_Orange_Slip.php
<?php
/**
 * Swiss Payment Slip FPDF
 *
 * @license http://www.opensource.org/licenses/mit-license.php MIT License
 * @copyright 2012-2015 Some nice Swiss guys
 * @author Marc Würth <ravage@bluewin.ch>
 * @author Manuel Reinhard <manu@sprain.ch>
 * @author Peter Siska <pesche@gridonic.ch>
 * @link https://github.com/ravage84/SwissPaymentSlipFpdf
 */
?>
<!DOCTYPE html>
<html>
<head>
    <meta charset="utf-8">
    <title>SwissPaymentSlipTcpdf Example 01: SwissPaymentSlipFpdf Create an orange payment slip</title>
</head>
<body>
<h1>SwissPaymentSlipFpdf Example 01: Create an orange payment slip</h1>
<?php
// Measure script execution/generating time
$time_start = microtime(true);

// Make sure the classes get auto-loaded
require __DIR__ . '/../vendor/autoload.php';

// Import necessary classes
use SwissPaymentSlip\SwissPaymentSlip\OrangePaymentSlipData;
use SwissPaymentSlip\SwissPaymentSlip\OrangePaymentSlip;
use SwissPaymentSlip\SwissPaymentSlipFpdf\PaymentSlipFpdf;
use fpdf\FPDF;

// Make sure FPDF has access to the additional fonts
define('FPDF_FONTPATH', __DIR__ . '/../src/Resources/font');

// Create an instance of FPDF, setup default settings
$fPdf = new FPDF('P', 'mm', 'A4');

// Add OCRB font to FPDF
$fPdf->AddFont('OCRB10');

// Add page, don't break page automatically
$fPdf->AddPage();
$fPdf->SetAutoPageBreak(false);

// Insert a dummy invoice text, not part of the payment slip itself
$fPdf->SetFont('Helvetica', '', 9);
$fPdf->Cell(50, 4, "Just some dummy text.");

// Create a payment slip data container (value object)
$paymentSlipData = new OrangePaymentSlipData();

// Fill the data container with your data
$paymentSlipData->setBankData('Seldwyla Bank', '8001 Zürich');
$paymentSlipData->setAccountNumber('01-145-6');
$paymentSlipData->setRecipientData('H. Muster AG', 'Versandhaus', 'Industriestrasse 88', '8000 Zürich');
$paymentSlipData->setPayerData('Rutschmann Pia', 'Marktgasse 28', '9400 Rorschach');
$paymentSlipData->setAmount(2830.50);
$paymentSlipData->setReferenceNumber('7520033455900012');
$paymentSlipData->setBankingCustomerId('215703');

// Create a payment slip object, pass in the prepared data container
$paymentSlip = new OrangePaymentSlip($paymentSlipData, 0, 191);

// Create an instance of the FPDF implementation
$paymentSlipFpdf = new PaymentSlipFpdf($fPdf, $paymentSlip);

// "Print" the slip with its elements according to their attributes
$paymentSlipFpdf->createPaymentSlip($paymentSlip);

// Output PDF named example_fdf_orange_slip.pdf to examples folder
$pdfName = 'example_fpdf_orange_slip.pdf';
$pdfPath = __DIR__ . DIRECTORY_SEPARATOR . $pdfName;
$fPdf->Output($pdfPath, 'F');

echo sprintf('Payment slip created in <a href="%s">%s</a><br>', $pdfName, $pdfPath);

echo "<br>";

$time_end = microtime(true);
$time = $time_end - $time_start;
echo "Generation took $time seconds <br>";
echo 'Peak memory usage: ' . memory_get_peak_usage() / 1024 / 1024;
?>
</body>
</html>


src/SwissPaymentSlip/SwissPaymentSlipFpdf/PaymentSlipFpdf.php
<?php
/**
 * Swiss Payment Slip FPDF
 *
 * @license http://www.opensource.org/licenses/mit-license.php MIT License
 * @copyright 2012-2015 Some nice Swiss guys
 * @author Marc Würth <ravage@bluewin.ch>
 * @author Manuel Reinhard <manu@sprain.ch>
 * @author Peter Siska <pesche@gridonic.ch>
 * @link https://github.com/ravage84/SwissPaymentSlipFpdf
 */

namespace SwissPaymentSlip\SwissPaymentSlipFpdf;

use SwissPaymentSlip\SwissPaymentSlipPdf\PaymentSlipPdf;
use fpdf\FPDF;

/**
 * Create Swiss payment slips (ESR/ES) as PDFs with FPDF
 *
 * Responsible for generating standard Swiss payment Slips using FPDF as engine.
 * Layouting done by utilizing SwissPaymentSlip
 * Data organisation through SwissPaymentSlipData
 */
class PaymentSlipFpdf extends PaymentSlipPdf {
    protected $rgbColors = array();

    /**
     * The PDF engine object to generate the PDF output with
     *
     * @var null|FPDF The PDF engine object
     */
    protected $pdfEngine = null;

    /**
     * The last set font family, prevents the PDF engine to re-set the same values over and over
     *
     * @var string;
     */
    protected $lastFontFamily = '';

    /**
     * The last set font size, prevents the PDF engine to re-set the same values over and over
     *
     * @var int|double;
     */
    protected $lastFontSize = '';

    /**
     * The last set font color, prevents the PDF engine to re-set the same values over and over
     *
     * @var string;
     */
    protected $lastFontColor = '';

    /**
     * {@inheritDoc}
     */
    protected function displayImage($background) {
        // TODO check if slipBackground is a color or a path to a file

        $this->pdfEngine->Image($background,
            $this->paymentSlip->getSlipPosX(),
            $this->paymentSlip->getSlipPosY(),
            $this->paymentSlip->getSlipWidth(),
            $this->paymentSlip->getSlipHeight(),
            strtoupper(substr($background, -3, 3)));
    }

    /**
     * {@inheritDoc}
     */
    protected function setFont($fontFamily, $fontSize, $fontColor) {
        if ($fontColor) {
            if ($this->lastFontColor != $fontColor) {
                $this->lastFontColor = $fontColor;

                $rgbArray = $this->convertColor2Rgb($fontColor);
                $this->pdfEngine->SetTextColor($rgbArray['red'], $rgbArray['green'], $rgbArray['blue']);
            }
        }
        if ($this->lastFontFamily != $fontFamily || $this->lastFontSize != $fontSize) {
            $this->lastFontFamily = $fontFamily;
            $this->lastFontSize = $fontSize;

            $this->pdfEngine->SetFont($fontFamily, '', $fontSize);
        }
    }

    /**
     * {@inheritDoc}
     */
    protected function setBackground($background) {
        // TODO check if it's a path to a file
        // TODO else it should be a color
        $rgbArray = $this->convertColor2Rgb($background);
        $this->pdfEngine->SetFillColor($rgbArray['red'], $rgbArray['green'], $rgbArray['blue']);
    }

    /**
     * {@inheritDoc}
     */
    protected function setPosition($posX, $posY) {
        $this->pdfEngine->SetXY($posX, $posY);
    }

    /**
     * {@inheritDoc}
     */
    protected function createCell($width, $height, $line, $textAlign, $fill) {
        $this->pdfEngine->Cell($width, $height, utf8_decode($line), 0, 0, $textAlign, $fill);
    }

    /**
     * {@inheritDoc}
     */
    protected function convertColor2Rgb($color) {
        if (isset($this->rgbColors[$color])) {
            return $this->rgbColors[$color];
        }
        $this->rgbColors[$color] = $this->hex2RGB($color);
        return $this->rgbColors[$color];
    }

    /**
     * Convert hexadecimal color code into an associative array or string of RGB values
     *
     * @param string $hexStr The hexadecimal color code (3 or 6 characters long)
     * @param bool $returnAsString Whether to return as a string or array.
     * @param string $separator The separator for the RGB value string, defaults to ",".
     * @return array|string|false The RGB values as an associative array or a string.
     * False if an invalid color code was given.
     *
     * @copyright 2010 hafees at msn dot com
     * @link http://www.php.net/manual/en/function.hexdec.php#99478
     * @todo Throw an exception if an invalid hex color code was given
     */
    private function hex2RGB($hexStr, $returnAsString = false, $separator = ',') {
        $hexStr = preg_replace("/[^0-9A-Fa-f]/", '', $hexStr); // Gets a proper hex string
        $rgbArray = array();
        if (strlen($hexStr) == 6) { //If a proper hex code, convert using bitwise operation. No overhead... faster
            $colorVal = hexdec($hexStr);
            $rgbArray['red'] = 0xFF & ($colorVal >> 0x10);
            $rgbArray['green'] = 0xFF & ($colorVal >> 0x8);
            $rgbArray['blue'] = 0xFF & $colorVal;
        } elseif (strlen($hexStr) == 3) { //if shorthand notation, need some string manipulations
            $rgbArray['red'] = hexdec(str_repeat(substr($hexStr, 0, 1), 2));
            $rgbArray['green'] = hexdec(str_repeat(substr($hexStr, 1, 1), 2));
            $rgbArray['blue'] = hexdec(str_repeat(substr($hexStr, 2, 1), 2));
        } else {
            return false; //Invalid hex color code
        }
        return $returnAsString ? implode($separator, $rgbArray) : $rgbArray; // returns the rgb string or the associative array
    }
}


src/SwissPaymentSlip/SwissPaymentSlipFpdf/SwissPaymentSlipFpdf.php
<?php
/**
 * Swiss Payment Slip as PDF
 *
 * PHP version >= 5.3.0
 *
 * @licence MIT
 * @copyright 2012-2013 Some nice Swiss guys
 * @author Manuel Reinhard <manu@sprain.ch>
 * @author Peter Siska <pesche@gridonic.ch>
 * @author Marc Würth ravage@bluewin.ch
 * @link https://github.com/sprain/class.Einzahlungsschein.php
 * @version: 0.5.0
 */

namespace SwissPaymentSlip\SwissPaymentSlipFpdf;

use SwissPaymentSlip\SwissPaymentSlipPdf\PaymentSlipPdf;
use fpdf\FPDF;

/**
 * Responsible for generating standard Swiss payment Slips using FPDF as engine.
 * Layouting done by utilizing SwissPaymentSlip
 * Data organisation through SwissPaymentSlipData
 */
class SwissPaymentSlipFpdf extends PaymentSlipPdf
{
	protected $rgbColors = array();

	/**
	 * The PDF engine object to generate the PDF output
	 *
	 * @var null|FPDF The PDF engine object
	 */
	protected $pdfEngine = null;

	/**
	 * @var string;
	 */
	protected  $lastFontFamily = '';

	/**
	 * @var string;
	 */
	protected  $lastFontSize = '';

	/**
	 * @var string;
	 */
	protected  $lastFontColor = '';

	/**
	 * @param $background
	 * @return mixed|void
	 */
	protected function displayImage($background) {
		// TODO check if slipBackground is a color or a path to a file

		$this->pdfEngine->Image($background,
			$this->paymentSlip->getSlipPosX(),
			$this->paymentSlip->getSlipPosY(),
			$this->paymentSlip->getSlipWidth(),
			$this->paymentSlip->getSlipHeight(),
			strtoupper(substr($background, -3, 3)));
	}

	/**
	 * @param $fontFamily
	 * @param $fontSize
	 * @param $fontColor
	 * @return mixed|void
	 */
	protected function setFont($fontFamily, $fontSize, $fontColor) {
		if ($fontColor) {
			if ($this->lastFontColor != $fontColor) {
				$this->lastFontColor = $fontColor;

				$rgbArray = $this->convertColor2Rgb($fontColor);
				$this->pdfEngine->SetTextColor($rgbArray['red'], $rgbArray['green'], $rgbArray['blue']);
			}
		}
		if ($this->lastFontFamily != $fontFamily || $this->lastFontSize != $fontSize) {
			$this->lastFontFamily = $fontFamily;
			$this->lastFontSize = $fontSize;

			$this->pdfEngine->SetFont($fontFamily, '', $fontSize);
		}
	}

	/**
	 * @param $background
	 * @return mixed|void
	 */
	protected function setBackground($background) {
		// TODO check if it's a path to a file
		// TODO else it should be a color
		$rgbArray = $this->convertColor2Rgb($background);
		$this->pdfEngine->SetFillColor($rgbArray['red'], $rgbArray['green'], $rgbArray['blue']);
	}

	/**
	 * @param $posX
	 * @param $posY
	 * @return mixed|void
	 */
	protected function setPosition($posX, $posY) {
		$this->pdfEngine->SetXY($posX, $posY);
	}

	/**
	 * @param $width
	 * @param $height
	 * @param $line
	 * @param $textAlign
	 * @param $fill
	 * @return mixed|void
	 */
	protected function createCell($width, $height, $line, $textAlign, $fill) {
		$this->pdfEngine->Cell($width, $height, utf8_decode($line), 0, 0, $textAlign, $fill);
	}

	/**
	 * @param $color
	 * @return mixed
	 */
	protected function convertColor2Rgb($color) {
		if (isset($this->rgbColors[$color])) {
			return $this->rgbColors[$color];
		}
		$this->rgbColors[$color] = $this->hex2RGB($color);
		return $this->rgbColors[$color];
	}

	/**
	 * Convert hexadecimal values into an array of RGB
	 *
	 * @param $hexStr
	 * @param bool $returnAsString
	 * @param string $separator
	 * @return array|bool|string
	 *
	 * @copyright 2010 hafees at msn dot com
	 * @link http://www.php.net/manual/en/function.hexdec.php#99478
	 */
	private function hex2RGB($hexStr, $returnAsString = false, $separator = ',') {
		$hexStr = preg_replace("/[^0-9A-Fa-f]/", '', $hexStr); // Gets a proper hex string
		$rgbArray = array();
		if (strlen($hexStr) == 6) { //If a proper hex code, convert using bitwise operation. No overhead... faster
			$colorVal = hexdec($hexStr);
			$rgbArray['red'] = 0xFF & ($colorVal >> 0x10);
			$rgbArray['green'] = 0xFF & ($colorVal >> 0x8);
			$rgbArray['blue'] = 0xFF & $colorVal;
		} elseif (strlen($hexStr) == 3) { //if shorthand notation, need some string manipulations
			$rgbArray['red'] = hexdec(str_repeat(substr($hexStr, 0, 1), 2));
			$rgbArray['green'] = hexdec(str_repeat(substr($hexStr, 1, 1), 2));
			$rgbArray['blue'] = hexdec(str_repeat(substr($hexStr, 2, 1), 2));
		} else {
			return false; //Invalid hex color code
		}
		return $returnAsString ? implode($separator, $rgbArray) : $rgbArray; // returns the rgb string or the associative array
	}
}


examples/SwissPaymentSlipFpdf_Thousand_Slips.php
<?php
/**
 * Swiss Payment Slip FPDF
 *
 * @license http://www.opensource.org/licenses/mit-license.php MIT License
 * @copyright 2012-2015 Some nice Swiss guys
 * @author Marc Würth <ravage@bluewin.ch>
 * @author Manuel Reinhard <manu@sprain.ch>
 * @author Peter Siska <pesche@gridonic.ch>
 * @link https://github.com/ravage84/SwissPaymentSlipFpdf
 */
?>
<!DOCTYPE html>
<html>
<head>
    <meta charset="utf-8">
    <title>SwissPaymentSlipFpdf Example 03: Create one thousand orange payment slips</title>
</head>
<body>
<h1>SwissPaymentSlipFpdf Example 03: Create one thousand orange payment slips</h1>
<?php
// Measure script execution/generating time
$time_start = microtime(true);

// Make sure the classes get auto-loaded
require __DIR__ . '/../vendor/autoload.php';

// Import necessary classes
use SwissPaymentSlip\SwissPaymentSlip\OrangePaymentSlipData;
use SwissPaymentSlip\SwissPaymentSlip\OrangePaymentSlip;
use SwissPaymentSlip\SwissPaymentSlipFpdf\PaymentSlipFpdf;
use fpdf\FPDF;

// Make sure FPDF has access to the additional fonts
define('FPDF_FONTPATH', __DIR__ . '/../src/Resources/font');

// Create an instance of FPDF, setup default settings
$fPdf = new FPDF('P', 'mm', 'A4');

// Add OCRB font to FPDF
$fPdf->AddFont('OCRB10');

// create 1000 payment slips
for ($slipNr = 1; $slipNr <= 1000; $slipNr++) {
    // Add page, don't break page automatically
    $fPdf->AddPage();
    $fPdf->SetAutoPageBreak(false);

    // Insert a dummy invoice text, not part of the payment slip itself
    $fPdf->SetFont('Helvetica', '', 9);
    $fPdf->Cell(50, 4, "Just some dummy text.");

    // Create a payment slip data container (value object)
    $paymentSlipData = new OrangePaymentSlipData();

    // Fill the data container with your data
    $paymentSlipData->setBankData('Seldwyla Bank', '8001 Zürich');
    $paymentSlipData->setAccountNumber('01-145-6');
    $paymentSlipData->setRecipientData('H. Muster AG', 'Versandhaus', 'Industriestrasse 88', '8000 Zürich');
    $paymentSlipData->setPayerData('Rutschmann Pia', 'Marktgasse 28', '9400 Rorschach', 'Slip # ' . $slipNr);
    $paymentSlipData->setAmount(2830.50);
    $paymentSlipData->setReferenceNumber('7520033455900012');
    $paymentSlipData->setBankingCustomerId('215703');

    // Create a payment slip object, pass in the prepared data container
    $paymentSlip = new OrangePaymentSlip($paymentSlipData, 0, 191);

    // Create an instance of the FPDF implementation
    $paymentSlipFpdf = new PaymentSlipFpdf($fPdf, $paymentSlip);

    // "Print" the slip with its elements according to their attributes
    $paymentSlipFpdf->createPaymentSlip($paymentSlip);
}

// Output PDF named example_fpdf_thousand_slips.pdf to examples folder
$pdfName = 'example_fpdf_thousand_slips.pdf';
$pdfPath = __DIR__ . DIRECTORY_SEPARATOR . $pdfName;
$fPdf->Output($pdfPath, 'F');

echo sprintf('Payment slips created in <a href="%s">%s</a><br>', $pdfName, $pdfPath);

echo "<br>";

$time_end = microtime(true);
$time = $time_end - $time_start;
echo "Generation took $time seconds <br>";
echo 'Peak memory usage: ' . memory_get_peak_usage() / 1024 / 1024;
?>
</body>
</html>
